### PR TITLE
Event for preparing anvil repairs. Fixes BUKKIT-2850

### PR DESCRIPTION
**Pull request by pruby**
Event to hook in to anvil activity. Allows the item to be created as a result of repair and the exp cost in levels to be changed.

https://github.com/Bukkit/CraftBukkit/issues/1030

### CODE REVIEW

**Comment by riking** [comment]
Can you `git rebase -i` and squash these into one?

**Comment by pruby** [comment]
Squashed in to one commit

**Comment by pruby** [comment]
Added documentation that result item stack may be null if nothing would normally be produced.

**Comment by riking** [comment]
Change the title - this is the event to handle anvils choosing the result.

**Comment by pruby** [comment]
Documented that this event can be triggered multiple times when an item is placed in the Javadoc. That seems unavoidable as the client sends multiple packets, each triggering that call when an item is initially placed.

**Comment by riking** [comment]
Oh yeah, do we have a test plugin to verify that anything changed actually displays right on the client?

**Comment by pruby** [comment]
My test was tacked on to another plugin I happened to be writing at the time - can pull it out in to a test plugin soon & get back to you.

**Comment by riking** [comment]
Well, I can take your word for it too. Just make sure to test that it shows up on the client.

**Comment by pruby** [comment]
Sorry, lost track and forgot to respond. Yes, changes here show up on the client.

**Comment by riking** [comment]
@gmcferrin added it to description, it seems he linked this one from the CB but not the other way around

File: src/main/java/org/bukkit/event/anvil/PrepareItemRepairEvent.java [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+
+    /**
+     * Set cost in exp levels of the repair.
+     * This may not show correctly on the client as the level cost is also calculated client-side.
+     * @param level - cost in levels
+     */
+    public void setExpLevelCost(int levelCost) {
+        this.levelCost = levelCost;
+    }
+
```
Comment (riking) [inline src/main/java/org/bukkit/event/anvil/PrepareItemRepairEvent.java:107]: Excess whitespace

File: src/main/java/org/bukkit/event/anvil/PrepareItemRepairEvent.java [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+        this.repairer = repairer;
+        this.anvil = anvil;
+        this.item1 = item1;
+        this.item2 = item2;
+        this.resultItem = resultItem;
+        this.levelCost = levelCost;
+        this.cancelled = false;
+    }
+    
+    public HumanEntity getRepairer() {
```
Comment (riking) [inline src/main/java/org/bukkit/event/anvil/PrepareItemRepairEvent.java:41]: Badly named

File: src/main/java/org/bukkit/event/anvil/PrepareItemRepairEvent.java [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+        super(view);
+        this.repairer = repairer;
+        this.anvil = anvil;
+        this.item1 = item1;
+        this.item2 = item2;
+        this.resultItem = resultItem;
+        this.levelCost = levelCost;
+        this.cancelled = false;
+    }
+
```
Comment (riking) [inline src/main/java/org/bukkit/event/anvil/PrepareItemRepairEvent.java:40]: Whitespace

File: src/main/java/org/bukkit/event/anvil/PrepareItemRepairEvent.java [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+    public HumanEntity getRepairer() {
+        return repairer;
+    }
+
+    /**
+     * Gets the anvil being used to repair the item
+     *
+     * @return the anvil block used
+     */
+    public Block getAnvilBlock() {
```
Comment (riking) [inline src/main/java/org/bukkit/event/anvil/PrepareItemRepairEvent.java:50]: Isn't there some kind of event where there's a getBlock()? Actually, what's the enchant prepare event do?

**Comment by gmcferrin** [comment]
From @amaranth:

> We wouldn't want to change the cost without that showing on the client so this just becomes a named replacement for InventoryClickEvent that erroneously gets called 3 times

Unless this is no longer true, we see no reason to pull this.

**Pull request closed without merging**